Analysis/src/Coda3EventDecoder.cc

```diff
@@ -0,0 +1,422 @@
+	buffer[0] = 4; // Go event length
+	buffer[1] = ((0xffd1 << 16) | (0x01 << 8) );
+	buffer[2] = localtime;
+	buffer[3] = 0; // unused
+	buffer[4] = eventcount;
+}
+
+void Coda3EventDecoder::EncodePauseEventHeader(int* buffer, int buffer_size)
+{
+	int localtime  = (int)time(0);
+	int eventcount = 0;
+	// TODO: We need access to the ControlEvent enum
+	buffer[0] = 4; // Pause event length
+  buffer[1] = ((0xffd2 << 16) | (0x01 << 8) );
+	buffer[2] =	localtime;
+	buffer[3] = 0; // unused
+	buffer[4] = eventcount;
+}
+
+void Coda3EventDecoder::EncodeEndEventHeader(int* buffer, int buffer_size)
+{
+	int localtime  = (int)time(0);
+	int eventcount = 0;
+	// TODO: We need access to the ControlEvent enum
+	buffer[0] = 4; // End event length
+  buffer[1] = ((0xffd3 << 16) | (0x01 << 8) );
+	buffer[2] = localtime;
+	buffer[3] = 0; // unused
+	buffer[4] = eventcount; 
+}
+
+
+Int_t Coda3EventDecoder::DecodeEventIDBank(UInt_t *buffer)
+{
+	// TODO:
+	// How should we handle bad events??
+  fPhysicsEventFlag = kFALSE;
+	fControlEventFlag = kFALSE;
+  Int_t ret = HED_OK;
+
+  // Main engine for decoding, called by public LoadEvent() methods
+  // this assert checks to see if buffer points to NULL
+  assert(buffer);
+	
+	// General Event information
+  fEvtLength = buffer[0]+1;  // in longwords (4 bytes)
+  fEvtType = 0;
+	fEvtTag = 0;
+  fBankDataType = 0;
+
+	// Prep TBOBJ variables
+	tbank.Clear();
+	tsEvType = 0;
+  evt_time = 0;
+	trigger_bits = 0;
+	block_size = 0;
+		
+	// Start Filling Data
+	fEvtTag     = (buffer[1] & 0xffff0000) >> 16;
+	fBankDataType = (buffer[1] & 0xff00) >> 8;
+	block_size  =	(buffer[1] & 0xff); 
+
+	if(block_size > 1) { QwWarning << "MultiBlock is not properly supported! block_size = " 
+											 				   << block_size << QwLog::endl; }
+	// Determine the event type by the call
+	fEvtType = InterpretBankTag(fEvtTag);		
+	fWordsSoFar = (2);
+	if(fEvtTag < 0xff00) { /* user event */ printUserEvent(buffer); }
+	else if(fControlEventFlag) {
+		fEvtNumber = 0;	ProcessControlEvent(fEvtType, &buffer[fWordsSoFar]);
+	}
+	else if(fPhysicsEventFlag) { 
+		ret = trigBankDecode( buffer );
+		if(ret != HED_OK) { trigBankErrorHandler( ret ); }
+		else { 
+			fEvtNumber = tbank.evtNum;
+			fWordsSoFar = 2 + tbank.len;
+		}
+	}
+	else { // Not a control event, user event, nor physics event. Not sure what it is
+      	 //  Arbitrarily set the event type to "fEvtTag".
+      	 //  The first two words have been examined.
+    QwWarning << "!!! Cannot determine the event type !!!" << QwLog::endl;
+		QwMessage << "Printing Event Buffer:";
+		QwMessage << "\n------------\n" << QwLog::endl;
+		for(size_t index = 0; fEvtLength; index++){
+ 			// TODO: // what if fEvtLength is gibberish because the event is gibberish?
+			QwMessage << "\t" << buffer[index];
+			if(index % 4 == 0){ QwMessage << QwLog::endl; }
+		}	
+		QwMessage << "\n------------\n" << QwLog::endl;
+
+ 		fEvtType = fEvtTag;	fEvtNumber = 0;
+	}
+
```

Comment:
Might want to change the order of the if-statements. PHYS events are the most common so they should be checked for first.